exec/cnex/format.c

```diff
@@ -0,0 +1,156 @@
+    spec->alternate = FALSE;
+    spec->sign = 0;
+    spec->width = -1;
+    spec->thousands = FALSE;
+    spec->precision = -1;
+    spec->type = 0;
+    if (string_isEmpty(fmt)) {
+        return TRUE;
+    }
+    size_t i = 0;
+    if (fmt->data[i] == '<' || fmt->data[i] == '>' || fmt->data[i] == '=' || fmt->data[i] == '^') {
+        spec->align = fmt->data[i];
+        ++i;
+    }
+    if (fmt->data[i] == '+' || fmt->data[i] == '-' || fmt->data[i] == ' ') {
+        spec->sign = fmt->data[i];
+        ++i;
+    }
+    if (fmt->data[i] == '0') {
+        spec->fill = '0';
+        if (spec->align == 0) {
+            spec->align = '=';
+        }
+        ++i;
+    }
+    if (isdigit(fmt->data[i])) {
+        size_t end = 0;
+        char *s = string_asCString(string_subString(fmt, i, fmt->length-1));
+        spec->width = atoi(s);
+        free(s);
+        i += end;
+    }
+    if (fmt->data[i] == ',') {
+        spec->thousands = TRUE;
+        ++i;
+    }
+    if (fmt->data[i] == '.') {
+        ++i;
+        size_t end = 0;
+        char *s = string_asCString(string_subString(fmt, i, fmt->length-1));
+        spec->precision = atoi(s);
+        free(s);
+        i += end;
+    }
+    if (i >= fmt->length - 1) {
+        spec->type = '\0';
+    } else {
+        spec->type = fmt->data[i];
+    }
+    switch (spec->type) {
+        case 0:
+        case 's':
+            if (spec->sign || spec->alternate || spec->align == '=') {
+                return FALSE;
+            }
+            break;
+        case 'd':
+        case 'x':
+            return TRUE;
+        default:
+            return FALSE;
+    }
+    return TRUE;
+}
+
+static TString *fillstr(Spec *spec, int n)
+{
+    TString *r = string_newString();
+    if (spec->fill < 128) {
+        return string_fillString(r, (char)spec->fill, n);
+    }
+
+    while (n > 0) {
+        r = string_appendChars(r, (char*)&spec->fill, 1);
+        n--;
+    }
+    return r;
+}
+
+TString *formatString(TExecutor *exec, TString *str, Spec *spec)
+{
+    static char hexdigit[] = "0123456789abcdef";
+
+    TString *r = str;
+    switch (spec->type) {
+        case 'd': {
+            char *s = string_asCString(str);
+            Number x = number_from_string(s);
+            free(s);
+            if (number_is_nan(x)) {
+                exec->rtl_raise(exec, "ValueRangeException", TCSTR(str), BID_ZERO);
+            }
+            break;
+        }
+        case 'x': {
+            char *s = string_asCString(str);
+            Number x = number_from_string(s);
+            free(s);
+            string_clearString(r);
+            if (number_is_zero(x)) {
+                r = string_appendChars(r, "0", 1);
+            } else {
+                Number base = number_from_uint32(16);
+                // TODO: handle negative numbers
+                while (!number_is_zero(x)) {
+                    r = string_appendChars(r, (char*)&hexdigit[number_to_uint32(number_modulo(x, base))], 1);
```

Comment:
No `(char*)` cast needed here. It's already a character pointer.

Reply:
Removed cast.